main.cc

```diff
@@ -594,20 +594,47 @@ void dump_performance_profiles() {
         __llvm_profile_dump();
     }
 }
+
 namespace {
+
+// Check whether raft based topology is enabled.
+future<bool> raft_topology_enabled(sharded<db::system_keyspace>& sys_ks) {
+    co_return service::string_to_group0_upgrade_state(sys_ks.local().load_group0_upgrade_state().get()) ==
+            service::group0_upgrade_state::use_post_raft_procedures;
+}
+
+// Get local node state from the system tables.
+std::optional<locator::node::state> get_local_node_state(sharded<db::system_keyspace>& sys_ks, const locator::host_id host_id) {
+    if (!raft_topology_enabled(sys_ks).get()) {
+        return std::nullopt;
+    }
+
+    std::optional<locator::node::state> node_state;
+    auto topo_state = sys_ks.local().load_topology_state({}).get();
+    const auto* res = topo_state.find(raft::server_id{host_id.id});
+    if (res) {
+        node_state = service::to_topology_node_state(res->second.state);
+    }
+    return node_state;
+}
+
+// Propagate the local node ID to topology.
 future<> add_local_node_to_topology(sharded<db::system_keyspace>& sys_ks, const locator::host_id host_id, const auto broadcast_addr,
         sharded<locator::shared_token_metadata>& token_metadata) {
-    locator::shared_token_metadata::mutate_on_all_shards(token_metadata, [host_id, endpoint = broadcast_addr](locator::token_metadata& tm) {
+    const auto node_state = get_local_node_state(sys_ks, host_id);
+
+    locator::shared_token_metadata::mutate_on_all_shards(token_metadata, [host_id, endpoint = broadcast_addr, node_state](locator::token_metadata& tm) {
         // Makes local host id available in topology cfg as soon as possible.
         // Raft topology discard the endpoint-to-id map, so the local id can
         // still be found in the config.
         tm.get_topology().set_host_id_cfg(host_id);
-        tm.get_topology().add_or_update_endpoint(host_id, endpoint);
+        tm.get_topology().add_or_update_endpoint(host_id, endpoint, std::nullopt, node_state);
```

Comment:
This patch does not insure that.  The ERM created at that point contains only the local node which is apparently required for further initialization. Then later when raft server starts it loads the topology from the system tables and a new ERM is created.

Reply:
Then I'm not sure what you mean in the commit log which says "This ensures correct initialization of ... data center list in topology". It probably needs to be revised.

Reply:
This change was removed since topology initialization during reboot is now done earlier and takes care of initializing the local node state in case the node was bootstrapped.